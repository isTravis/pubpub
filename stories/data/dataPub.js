export default {
	id: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
	slug: '6pewbpry',
	title: 'Defining the Dimensions of the “Space” of Computing',
	description:
		'The first computing machines were so large they filled entire rooms. Today they are ubiquitous, built invisibly into our environments. While it is tempting to view this change within a predetermined space of progress, we can still shape the future on our own terms.',
	avatar: 'https://assets.pubpub.org/9pi1c5c6/41554230047846.jpeg',
	useHeaderImage: true,
	firstPublishedAt: '2019-04-22T03:43:24.218Z',
	lastPublishedAt: '2019-04-22T03:43:24.218Z',
	draftEditHash: 'eyy8hltx',
	draftViewHash: 's5mjfpdz',
	doi: null,
	labels: null,
	isCommunityAdminManaged: true,
	review: null,
	downloads: null,
	communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
	createdAt: '2019-04-02T18:12:05.538Z',
	updatedAt: '2019-04-30T15:23:17.211Z',
	branches: [
		{
			createdAt: '2019-04-29T22:01:27.918Z',
			id: '93504929-a39b-4d97-a618-83e00dc666ca',
			shortId: 1,
			title: 'draft',
			description: null,
			submissionAlias: null,
			order: 0,
			communityAdminPermissions: 'none',
			publicPermissions: 'edit',
			permissions: [
				{
					id: '00ed4bc1-a408-4725-87f4-b07a41b2cfb5',
					permissions: 'edit',
					userId: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
					branchId: '93504929-a39b-4d97-a618-83e00dc666ca',
					pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
					createdAt: '2019-04-02T18:12:05.543Z',
					updatedAt: '2019-04-02T18:12:05.543Z',
					user: {
						id: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
						firstName: 'Travis',
						lastName: 'Cohen',
						fullName: 'Travis Cohen',
						avatar: null,
						slug: 'traviscohen',
						initials: 'TC',
						title: 'Travis',
					},
				},
			],
			isActive: true,
			viewHash: 'cat',
			discussHash: 'dog',
			editHash: 'fish',
		},
		{
			createdAt: '2019-04-29T22:01:27.919Z',
			id: 'ebd27c37-27ee-4c75-b907-d99263a8641a',
			shortId: 2,
			title: 'public',
			description: null,
			submissionAlias: 'Submit for Review',
			order: 1,
			communityAdminPermissions: 'none',
			publicPermissions: 'edit',
			permissions: [],
			isActive: false,
			viewHash: 'cat2',
			discussHash: 'dog2',
			editHash: 'fish2',
		},
	],
	managers: [
		{
			id: '00ed4bc1-a408-4725-87f4-b07a41b2cfb5',
			userId: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:05.543Z',
			updatedAt: '2019-04-02T18:12:05.543Z',
			user: {
				id: 'd6028d98-0cd9-4711-877b-bc5c3c7d6d4b',
				firstName: 'Travis',
				lastName: 'Cohen',
				fullName: 'Travis Cohen',
				avatar: null,
				slug: 'traviscohen',
				initials: 'TC',
				title: 'Travis',
			},
		},
		{
			id: '213eb90a-c6da-4617-8178-369b0a019cf1',
			userId: 'fc4d0b74-0fb0-413e-aea4-a9aecb3c1154',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:44:03.927Z',
			updatedAt: '2019-04-02T18:44:03.927Z',
			user: {
				id: 'fc4d0b74-0fb0-413e-aea4-a9aecb3c1154',
				firstName: 'Amy',
				lastName: 'Graves',
				fullName: 'Amy Graves',
				avatar: 'https://assets.pubpub.org/5k1fj9pu/51539111048243.jpg',
				slug: 'amy-graves',
				initials: 'AG',
				title: 'copy editor',
			},
		},
	],
	attributions: [
		{
			id: '146c7a07-43a7-49e1-8b04-734bbc7166c9',
			name: 'Weiwei Hsu',
			avatar: null,
			title: null,
			order: 0.25,
			isAuthor: true,
			roles: ['Writing – Original Draft Preparation'],
			userId: null,
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:48.718Z',
			updatedAt: '2019-04-02T18:12:53.295Z',
			user: {
				id: '146c7a07-43a7-49e1-8b04-734bbc7166c9',
				initials: 'W',
				fullName: 'Weiwei Hsu',
				firstName: 'Weiwei',
				lastName: 'Hsu',
				avatar: null,
				title: null,
			},
		},
		{
			id: '4f309ba3-5337-49f6-adee-8c921ae50d1d',
			name: 'Hugh Dubberly',
			avatar: null,
			title: null,
			order: 0.5,
			isAuthor: true,
			roles: ['Writing – Original Draft Preparation'],
			userId: null,
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			createdAt: '2019-04-02T18:12:43.574Z',
			updatedAt: '2019-04-02T18:12:56.267Z',
			user: {
				id: '4f309ba3-5337-49f6-adee-8c921ae50d1d',
				initials: 'H',
				fullName: 'Hugh Dubberly',
				firstName: 'Hugh',
				lastName: 'Dubberly',
				avatar: null,
				title: null,
			},
		},
	],
	collectionPubs: [
		{
			id: '88000923-6c19-401a-acf5-98effe5c9c5b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			collectionId: '65fc8561-31ea-4e2c-8067-b315dd7c933f',
			contextHint: null,
			rank: null,
			isPrimary: false,
			createdAt: '2019-04-24T20:43:15.375Z',
			updatedAt: '2019-04-24T20:43:15.375Z',
			collection: {
				id: '65fc8561-31ea-4e2c-8067-b315dd7c933f',
				title: '5. Essays in Exploration: Resisting Reduction',
				isRestricted: true,
				isPublic: true,
				pageId: 'f7c79073-8110-40c2-9d4f-9d090673cec0',
				communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
				metadata: null,
				kind: 'tag',
				doi: null,
				createdAt: '2019-01-22T03:16:13.644Z',
				updatedAt: '2019-01-22T03:17:58.495Z',
				page: {
					id: 'f7c79073-8110-40c2-9d4f-9d090673cec0',
					title: '5. Essays in Exploration: Resisting Reduction',
					slug: 'issue-5',
				},
			},
		},
		{
			id: '6cfea932-1701-4d4e-86b9-f85758dad57b',
			pubId: '00f9aaaf-0468-4590-9b86-1a2bff4ffe57',
			collectionId: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
			contextHint: null,
			rank: null,
			isPrimary: false,
			createdAt: '2019-04-24T20:43:15.375Z',
			updatedAt: '2019-04-24T20:43:15.375Z',
			collection: {
				id: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
				title: 'Home',
				isRestricted: true,
				isPublic: true,
				pageId: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
				communityId: '7808da6b-94d1-436d-ad79-2e036a8e4428',
				metadata: null,
				kind: 'tag',
				doi: null,
				createdAt: '2018-10-15T23:51:00.296Z',
				updatedAt: '2018-10-15T23:51:00.296Z',
				page: {
					id: '1e1c0f1f-1873-4705-b6e0-934f588889c6',
					title: 'Home',
					slug: '',
				},
			},
		},
	],
	discussions: [],
	activeBranch: {
		createdAt: '2019-04-29T22:01:27.918Z',
		id: '93504929-a39b-4d97-a618-83e00dc666ca',
		shortId: 1,
		title: 'draft',
		description: null,
		submissionAlias: null,
		order: 0,
		communityAdminPermissions: 'none',
		publicPermissions: 'edit',
		permissions: [],
		isActive: true,
		viewHash: 'cat2',
		discussHash: 'dog2',
		editHash: 'fish2',
	},
	isManager: true,
	isEditor: true,
	isViewer: false,
	isBranchAdmin: false,
	isStaticDoc: false,
	citationData: {
		pub: {
			apa:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry</div></div>',
			harvard:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry</div></div>',
			vancouver:
				'<div class="csl-bib-body">  <div data-csl-entry-id="00f9aaaf-0468-4590-9b86-1a2bff4ffe57" class="csl-entry">    <div class="csl-left-margin">1. </div><div class="csl-right-inline">Hsu W, Dubberly H. Defining the Dimensions of the “Space” of Computing. JoDS V6 Dev [Internet]. 2019 Apr 30; Available from: https://jods.mitpress.mit.edu/pub/6pewbpry</div>  </div></div>',
			bibtex:
				'<div class="csl-bib-body"><div class="csl-entry">@article{Hsu2019Defining,<ul style="list-style-type:none"><li>journal = {JoDS V6 Dev},</li><li>note = {https://jods.mitpress.mit.edu/pub/6pewbpry},</li><li>title = {Defining the Dimensions of the “Space” of Computing},</li><li>url = {https://jods.mitpress.mit.edu/pub/6pewbpry},</li><li>author = {Hsu, Weiwei and Dubberly, Hugh},</li><li>date = {2019-04-30},</li><li>year = {2019},</li><li>month = {4},</li><li>day = {30},</li></ul>}</div></div>',
		},
		version: {
			apa:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div></div>',
			harvard:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">Hsu, W., &#38; Dubberly, H. (2019). Defining the Dimensions of the “Space” of Computing. <i>JoDS V6 Dev</i>. Retrieved from https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div></div>',
			vancouver:
				'<div class="csl-bib-body">  <div data-csl-entry-id="Draft" class="csl-entry">    <div class="csl-left-margin">1. </div><div class="csl-right-inline">Hsu W, Dubberly H. Defining the Dimensions of the “Space” of Computing. JoDS V6 Dev [Internet]. 2019 Apr 30; Available from: https://jods.mitpress.mit.edu/pub/6pewbpry/draft</div>  </div></div>',
			bibtex:
				'<div class="csl-bib-body"><div class="csl-entry">@article{Hsu2019Defining,<ul style="list-style-type:none"><li>journal = {JoDS V6 Dev},</li><li>note = {https://jods.mitpress.mit.edu/pub/6pewbpry/draft},</li><li>title = {Defining the Dimensions of the “Space” of Computing},</li><li>url = {https://jods.mitpress.mit.edu/pub/6pewbpry/draft},</li><li>author = {Hsu, Weiwei and Dubberly, Hugh},</li><li>date = {2019-04-30},</li><li>year = {2019},</li><li>month = {4},</li><li>day = {30},</li></ul>}</div></div>',
		},
	},
	firebaseToken:
		'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJodHRwczovL2lkZW50aXR5dG9vbGtpdC5nb29nbGVhcGlzLmNvbS9nb29nbGUuaWRlbnRpdHkuaWRlbnRpdHl0b29sa2l0LnYxLklkZW50aXR5VG9vbGtpdCIsImlhdCI6MTU1NjYzODU3NCwiZXhwIjoxNTU2NjQyMTc0LCJpc3MiOiJmaXJlYmFzZS1hZG1pbnNkay1pZTd5ZUBwdWJwdWItdjYuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLCJzdWIiOiJmaXJlYmFzZS1hZG1pbnNkay1pZTd5ZUBwdWJwdWItdjYuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLCJ1aWQiOiJiMjQyZjYxNi03YWFhLTQ3OWMtOGVlNS0zOTMzZGNmNzA4NTkifQ.AFdYAfq3y1F1lgLfC0HKvfg4ZkqOP6draEEHBVDJM1ioxh8N9guJ6PM9rieoKry2UxNRCY3t9BttYXIffotOJLjj3L6MKq0FaOoEBycPZcrhqEQX43oTna54Zt54WD2ZKqGKO0TT1iUAdxJLJfqfthKe_o7IkXlc27z7uu09ek2RGBAvuiS10Jzc4j5urNQuLh6UFEVdoMaxBollkjEwK2wV333Df4TZ9jyfZCw7lR8VTPyXoXmVJbMUccLNM-i2tNO6FEqJ-C132qL19uGcw4LQtjwzSV0MQJJAhs16ZeEKPTkDF-XuOL-41ycq2tH0DBUlhHjAlhgbs53hCynVRg',
	mode: 'manage',
};
