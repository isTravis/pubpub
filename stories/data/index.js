export const accentDataDark = {
	accentColor: '#D13232',
	accentTextColor: '#FFF',
	accentActionColor: '#A72828',
	accentHoverColor: '#BC2D2D',
	accentMinimalColor: 'rgba(209, 50, 50, 0.15)',
};

export const accentDataLight = {
	accentColor: '#26E0D0',
	accentTextColor: '#000',
	accentActionColor: '#51E6D9',
	accentHoverColor: '#3BE3D4',
	accentMinimalColor: 'rgba(38, 224, 208, 0.15)',
};

import initialData from './dataInitial.js';
export const { locationData, loginData } = initialData;

export { default as communityData } from './dataCommunity.js';
export { default as collectionData } from './dataCollection.js';
export { default as pubData } from './dataPub.js';
export { plainDoc, imageDoc, fullDoc } from './dataDocs.js';
