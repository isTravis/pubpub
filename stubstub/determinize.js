export const determinize = (deterministicKeys) => (object) => {
	const res = {};
	Object.keys(object).forEach((key) => {
		if (deterministicKeys.includes(key)) {
			res[key] = object[key];
		}
	});
	return res;
};
