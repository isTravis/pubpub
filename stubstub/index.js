export { setup, teardown } from './prepare';
export { makeUser, makeCommunity } from './modelHelpers';
export { login } from './userToAgentMap';
export { stubModule as stub, stubFirebaseAdmin } from './stub';
