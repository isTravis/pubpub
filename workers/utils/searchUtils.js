import stopword from 'stopword';
import stopWordList from './stopwords';
import {
	Pub,
	Community,
	Branch,
	BranchPermission,
	PubAttribution,
	User,
	PubManager,
	Page,
} from '../../server/models';
import { getBranchDoc } from '../../server/utils/firebaseAdmin';

const lengthInUtf8Bytes = (str) => {
	// Matches only the 10.. bytes that are non-initial characters in a multi-byte sequence.
	const m = encodeURIComponent(str).match(/%[89ABab]/g);
	return str.length + (m ? m.length : 0);
};

const validateTextSize = (stringArray) => {
	return stringArray
		.map((string) => {
			const stringSize = lengthInUtf8Bytes(string);
			if (stringSize > 8500) {
				const re = new RegExp(
					`.{1,${Math.floor(8000 / (Math.ceil(stringSize / 8000) + 1))}}`,
					'g',
				);
				return string.match(re);
			}
			return string;
		})
		.reduce((prev, curr) => {
			/* Flatten array */
			return prev.concat(curr);
		}, []);
};

const jsonToTextChunks = (docJson) => {
	const getText = (node) => {
		if (node.content) {
			return node.content.reduce((prev, curr) => {
				return `${prev} ${getText(curr)}`;
			}, '');
		}
		if (node.text) {
			return node.text;
		}
		return '';
	};
	const text = typeof docJson === 'string' ? '' : getText(docJson).replace(/\s\s+/g, ' ');
	const splitText = stopword
		.removeStopwords(text.split(' '), stopWordList)
		.join(' ')
		.match(/.{1,8000}/g) || [''];
	return validateTextSize(splitText);
};

export const getPubSearchData = (pubIds) => {
	const dataToSync = [];
	return Pub.findAll({
		where: {
			id: pubIds,
		},
		include: [
			{
				model: Community,
				as: 'community',
			},
			{
				model: PubAttribution,
				as: 'attributions',
				required: false,
				separate: true,
				include: [
					{
						model: User,
						as: 'user',
						required: false,
						attributes: ['id', 'fullName'],
					},
				],
			},
			{
				model: Branch,
				as: 'branches',
				required: true,
				include: [
					{
						model: BranchPermission,
						as: 'permissions',
						separate: true,
						required: false,
					},
				],
			},
			{
				model: PubManager,
				as: 'managers',
				separate: true,
			},
		],
	})
		.then((pubs) => {
			const branchesToSync = [];
			pubs.forEach((pubData) => {
				const pub = pubData.toJSON();
				const authorByline = pub.attributions
					.map((attribution) => {
						if (attribution.user) {
							return attribution;
						}
						return {
							...attribution,
							user: {
								id: attribution.id,
								fullName: attribution.name,
							},
						};
					})
					.filter((attribution) => {
						return attribution.isAuthor;
					})
					.sort((foo, bar) => {
						if (foo.order < bar.order) {
							return -1;
						}
						if (foo.order > bar.order) {
							return 1;
						}
						if (foo.createdAt < bar.createdAt) {
							return 1;
						}
						if (foo.createdAt > bar.createdAt) {
							return -1;
						}
						return 0;
					})
					.map((author, index, array) => {
						const separator =
							index === array.length - 1 || array.length === 2 ? '' : ', ';
						const prefix = index === array.length - 1 && index !== 0 ? ' and ' : '';
						return `${prefix}${author.user.fullName}${separator}`;
					})
					.join('');

				const managerIds = pub.managers.map((manager) => {
					return manager.userId;
				});

				pub.branches.forEach((branch) => {
					const branchPermissionIds = branch.permissions.map((branchPermission) => {
						return branchPermission.userId;
					});
					/* Assume metadata is 2000 characters = 2000 bytes */
					const data = {
						pubId: pub.id,
						title: pub.title,
						slug: pub.slug,
						avatar: pub.avatar,
						description: pub.description,
						byline: authorByline ? `by ${authorByline}` : '',
						communityId: pub.community.id,
						communityDomain:
							pub.community.domain || `${pub.community.subdomain}.pubpub.org`,
						communityTitle: pub.community.title,
						communityAccentColorLight: pub.community.accentColorLight,
						communityAccentColorDark: pub.community.accentColorDark,
						communityHeaderLogo: pub.community.headerLogo,
						communityHeaderColorType: pub.community.headerColorType,
						communityUseHeaderTextAccent: pub.community.useHeaderTextAccent,
						branchId: branch.id,
						branchShortId: branch.shortId,
						branchIsPublic: branch.publicPermissions !== 'none',
						branchAdminAccessId:
							branch.communityAdminPermissions !== 'none' && pub.community.id,
						branchAccessIds: [...branchPermissionIds, ...managerIds],
						branchCreatedAt: branch.createdAt,
						branchContent: undefined,
					};
					branchesToSync.push(data);
				});
			});
			return branchesToSync;
		})
		.then((branchesToSync) => {
			return branchesToSync.reduce((promise, branchData) => {
				return promise
					.then(() => {
						return getBranchDoc(branchData.pubId, branchData.branchId);
					})
					.then((branchDocData) => {
						const { content } = branchDocData;
						if (content) {
							jsonToTextChunks(content).forEach((textChunk) => {
								dataToSync.push({
									...branchData,
									branchContent: textChunk,
								});
							});
						}
					});
			}, Promise.resolve());
		})
		.then(() => {
			return dataToSync;
		});
};

export const getPageSearchData = (pageIds) => {
	return Page.findAll({
		where: {
			id: pageIds,
		},
		include: [
			{
				model: Community,
				as: 'community',
			},
		],
	}).then((pages) => {
		const dataToSync = [];
		pages.forEach((page) => {
			const data = {
				pageId: page.id,
				title: page.title,
				slug: page.slug,
				avatar: page.avatar,
				description: page.description,
				isPublic: page.isPublic,
				communityId: page.community.id,
				communityDomain: page.community.domain || `${page.community.subdomain}.pubpub.org`,
				communityTitle: page.community.title,
				communityAccentColorLight: page.community.accentColorLight,
				communityAccentColorDark: page.community.accentColorDark,
				communityHeaderLogo: page.community.headerLogo,
				communityHeaderColorType: page.community.headerColorType,
				communityUseHeaderTextAccent: page.community.useHeaderTextAccent,
				content: undefined,
			};

			const layout = page.layout || [];
			let hasContent = false;
			layout
				.filter((block) => {
					return block.type === 'text' && block.content.text;
				})
				.forEach((block) => {
					jsonToTextChunks(block.content.text).forEach((textChunk) => {
						hasContent = true;
						dataToSync.push({
							...data,
							content: textChunk,
						});
					});
				});
			if (!hasContent) {
				dataToSync.push(data);
			}
		});
		return dataToSync;
	});
};
