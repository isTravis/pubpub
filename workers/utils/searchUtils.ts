import chunkText from 'chunk-text';

import {
	Pub,
	Community,
	PubAttribution,
	Page,
	Release,
	Doc,
	includeUserModel,
} from 'server/models';
import { jsonToNode } from 'components/Editor';
import { getPubDraftDoc } from 'server/utils/firebaseAdmin';
import { getScope, getMembers } from 'server/utils/queryHelpers';
import { getAuthorString } from 'utils/contributors';
import { DefinitelyHas, Release as ReleaseType, Pub as PubType, DocJson } from 'types';

import { stopwords } from './stopwords';

type AlgoliaPubEntry = {
	pubId: string;
	title: string;
	slug: string;
	avatar: string;
	description: string;
	byline: string;
	customPublishedAt: string;
	communityId: string;
	communityDomain: string;
	communityTitle: string;
	communityAccentColorLight: string;
	communityAccentColorDark: string;
	communityHeaderLogo: string;
	communityHeaderColorType: string;
	communityUseHeaderTextAccent: boolean;
	userIdsWithAccess: string;
	isPublic: boolean;
	content: string | string[]; // Currently just `string` but historically either
} & ({ isPublic: false; userIdsWithAccess: string[] } | { isPublic: true });

type SearchPub = DefinitelyHas<PubType, 'attributions' | 'community'> & {
	releases: DefinitelyHas<ReleaseType, 'doc'>[];
};

const docJsonToTextChunks = (docJson: DocJson): string[] => {
	const { textContent } = jsonToNode(docJson);
	const contentWithoutStopwords = textContent
		.split(' ')
		.filter((x) => !stopwords.has(x))
		.join(' ');
	return chunkText(contentWithoutStopwords, 7500, {
		// Count each byte in every character
		charLengthMask: 0,
	});
};

const createSearchDataForPub = async (pub: SearchPub): Promise<AlgoliaPubEntry[]> => {
	const {
		releases: [release],
		community,
	} = pub;
	const entries: AlgoliaPubEntry[] = [];
	const scopeData = await getScope({ pubId: pub.id, communityId: pub.community.id });
	const { members } = await getMembers({ scopeData });
	const userIdsWithAccess = members.map((m) => m.userId);
	const authorByline = getAuthorString(pub);
	const sharedFields = {
		pubId: pub.id,
		title: pub.title,
		slug: pub.slug,
		avatar: pub.avatar!,
		description: pub.description!,
		byline: authorByline ? `by ${authorByline}` : '',
		customPublishedAt: pub.customPublishedAt!,
		communityId: community.id,
		communityDomain: community.domain || `${community.subdomain}.pubpub.org`,
		communityTitle: community.title,
		communityAccentColorLight: community.accentColorLight!,
		communityAccentColorDark: community.accentColorDark!,
		communityHeaderLogo: community.headerLogo!,
		communityHeaderColorType: community.headerColorType!,
		communityUseHeaderTextAccent: community.useHeaderTextAccent!,
		userIdsWithAccess,
	};
	if (release) {
		const releaseContentChunks = docJsonToTextChunks(release.doc.content);
		releaseContentChunks.forEach((chunk) =>
			entries.push({
				...sharedFields,
				isPublic: true,
				content: chunk,
			}),
		);
	}
	const { doc } = await getPubDraftDoc(pub.id);
	const draftContentChunks = docJsonToTextChunks(doc);
	draftContentChunks.forEach((chunk) =>
		entries.push({
			...sharedFields,
			isPublic: false,
			userIdsWithAccess,
			content: chunk,
		}),
	);
	return entries;
};

export const getPubSearchData = async (pubIds) => {
	const pubs: SearchPub[] = await Pub.findAll({
		where: { id: pubIds },
		include: [
			{
				model: Community,
				as: 'community',
			},
			{
				model: PubAttribution,
				as: 'attributions',
				required: false,
				separate: true,
				include: [includeUserModel({ as: 'user' })],
			},
			{
				model: Release,
				as: 'releases',
				separate: true,
				include: [{ model: Doc, as: 'doc' }],
				order: [['historyKey', 'DESC']],
				limit: 1,
			},
		],
	});
	const resultsByPub = await Promise.all(pubs.map(createSearchDataForPub));
	return resultsByPub.reduce((a, b) => [...a, ...b], []);
};

export const getPageSearchData = async (pageIds) => {
	const pages = await Page.findAll({
		where: {
			id: pageIds,
		},
		include: [
			{
				model: Community,
				as: 'community',
			},
		],
	});
	const dataToSync: any[] = [];
	for (let index = 0; index < pages.length; index++) {
		const page = pages[index];
		// eslint-disable-next-line no-await-in-loop
		const scopeData = await getScope({ communityId: page.community.id });
		// eslint-disable-next-line no-await-in-loop
		const { members } = await getMembers({ scopeData });
		const accessIds = members.map((member) => {
			return member.userId;
		});
		const data = {
			pageId: page.id,
			title: page.title,
			slug: page.slug,
			avatar: page.avatar,
			description: page.description,
			isPublic: page.isPublic,
			communityId: page.community.id,
			communityDomain: page.community.domain || `${page.community.subdomain}.pubpub.org`,
			communityTitle: page.community.title,
			communityAccentColorLight: page.community.accentColorLight,
			communityAccentColorDark: page.community.accentColorDark,
			communityHeaderLogo: page.community.headerLogo,
			communityHeaderColorType: page.community.headerColorType,
			communityUseHeaderTextAccent: page.community.useHeaderTextAccent,
			pageAccessIds: accessIds,
			content: undefined,
		};

		const layout = page.layout || [];
		let hasContent = false;
		layout
			.filter((block) => {
				return block.type === 'text' && block.content.text;
			})
			.forEach((block) => {
				docJsonToTextChunks(block.content.text).forEach((textChunk) => {
					hasContent = true;
					dataToSync.push({
						...data,
						content: textChunk,
					});
				});
			});
		if (!hasContent) {
			dataToSync.push(data);
		}
	}

	return dataToSync;
};
