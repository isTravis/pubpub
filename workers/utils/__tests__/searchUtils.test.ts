import { stub, modelize, setup, teardown, determinize } from 'stubstub';
import { plainDoc, fullDoc, imageDoc } from 'utils/storybook/data';

import { getPageSearchData, getPubSearchData } from '../searchUtils';

const models = modelize`
    Community {
        title: "Test Community"
        subdomain: "tc1"
        Pub draftPub {
            title: "A Draft Pub"
            slug: "a-draft-pub"
            description: "You probably cannot see me"
            avatar: "shh.png"
            PubAttribution {
                User {
                    slug: "test1"
                }
            }
        }
        Pub releasedPub {
            title: "A Released Pub"
            slug: "a-released-pub"
            description: "Hey look at me"
            avatar: "hello.png"
            PubAttribution {
                isAuthor: false
                order: 0.3
                User {
                    slug: "test3"
                }
            }
            PubAttribution {
                isAuthor: true
                order: 0.2
                User {
                    firstName: "First"
                    lastName: "Author"
                    fullName: "Second Author"
                    slug: "test4"
                }
            }
            PubAttribution {
                isAuthor: true
                order: 0.1
                User {
                    firstName: "First"
                    lastName: "Author"
                    fullName: "First Author"
                    slug: "test2"
                }
            }
            Release {
								User {}
                createdAt: "2021-01-01"
                historyKey: 1
                Doc {
                    content: ${plainDoc}
                }
            }
            Release {
								User {}
                createdAt: "2020-01-02"
                historyKey: 2
                Doc {
                    content: ${fullDoc}
                }
            }
        }
        Page testPage {
            title: "I am a test page"
            slug: "test-page"
            description: "You can put text or Pubs on me"
            avatar: "page.png"
        }
    }
`;

let firebaseStub;

setup(beforeAll, async () => {
	await models.resolve();
	firebaseStub = stub('server/utils/firebaseAdmin', {
		getPubDraftDoc: () => {
			return Promise.resolve({ doc: imageDoc });
		},
	});
});

teardown(afterAll, () => {
	firebaseStub.restore();
});

const determinizePubData = determinize([
	'avatar',
	'byline',
	'communityDomain',
	'communityTitle',
	'description',
	'slug',
	'title',
	'content',
	'isPublic',
]);

const determinizePageData = determinize([
	'avatar',
	'communityDomain',
	'communityTitle',
	'description',
	'slug',
	'title',
]);

describe('getPubSearchData', () => {
	it('produces the expected data for a draft Pub', async () => {
		const data = await getPubSearchData(models.draftPub.id);
		expect(data.map(determinizePubData)).toMatchSnapshot();
	});

	it('produces the expected data for a released Pub', async () => {
		const data = await getPubSearchData(models.releasedPub.id);
		expect(data.map(determinizePubData)).toMatchSnapshot();
	});
});

describe('getPageSearchData', () => {
	it('produces the expected data for a Page', async () => {
		const data = await getPageSearchData(models.testPage.id);
		expect(data.map(determinizePageData)).toMatchSnapshot();
	});
});
