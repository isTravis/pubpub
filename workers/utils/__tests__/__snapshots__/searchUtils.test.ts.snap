// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPageSearchData produces the expected data for a Page 1`] = `
Array [
  Object {
    "avatar": "page.png",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "You can put text or Pubs on me",
    "slug": "test-page",
    "title": "I am a test page",
  },
]
`;

exports[`getPubSearchData produces the expected data for a draft Pub 1`] = `
Array [
  Object {
    "avatar": "shh.png",
    "byline": "by Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "content": "Hello hello.IntroductionWhateverOkay",
    "description": "You probably cannot see me",
    "isPublic": false,
    "slug": "a-draft-pub",
    "title": "A Draft Pub",
  },
]
`;

exports[`getPubSearchData produces the expected data for a released Pub 1`] = `
Array [
  Object {
    "avatar": "hello.png",
    "byline": "by First Author, Second Author, and Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "content": "IntroductionWelcome introduction!This bold I pretty nifty . Nature’s ecosystem elegant example complex adaptive system myriad “currencies” interact respond feedback systems enable flourishing regulation. This collaborative model–rather model exponential financial growth Singularity, promises transcendence current human condition advances technology—should provide paradigm approach artificial intelligence. More 60 years ago, MIT mathematician philosopher Norbert Wiener warned “when human atoms knit organization used, full responsible human beings, cogs levers rods, matters raw material flesh blood.” We heed Wiener’s warning.As currencies abundant output successful process organism, organisms evolved output convert else. Over billions years, Earth’s ecosystem evolved, creating vast systems metabolic pathways forming highly complex self-regulating systems that, example, stabilize body temperatures temperature Earth, despite continuous fluctuations changes individual elements scale—from micro macro. The output process input another. Ultimately, interconnects.WhateverVideo SectionHelloAnd paragraph text Hello, you?Citation ListHelloAnd paragraph text Hello, you?InfoContactCountryAlfreds FutterkisteMaria AndersGermanyCentro comercial MoctezumaFrancisco ChangMexico",
    "description": "Hey look at me",
    "isPublic": true,
    "slug": "a-released-pub",
    "title": "A Released Pub",
  },
  Object {
    "avatar": "hello.png",
    "byline": "by First Author, Second Author, and Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "content": "Hello hello.IntroductionWhateverOkay",
    "description": "Hey look at me",
    "isPublic": false,
    "slug": "a-released-pub",
    "title": "A Released Pub",
  },
]
`;
